pkg/controller/poolcoordinator/poolcoordinator_controller.go

```diff
@@ -0,0 +1,378 @@
+	klog.Infof("ctl: %v", ctl)
+	return ctl
+}
+
+func (pcc *Controller) taintNodeNotSchedulable(name string) {
+	node, err := pcc.nodeLister.Get(name)
+	if err != nil {
+		klog.Error(err)
+		return
+	}
+	taints := node.Spec.Taints
+	if utils.TaintKeyExists(taints, constant.NodeNotSchedulableTaint) {
+		klog.Infof("taint %s: key %s already exists, nothing to do\n", name, constant.NodeNotSchedulableTaint)
+		return
+	}
+	nn := node.DeepCopy()
+	t := corev1.Taint{
+		Key:    constant.NodeNotSchedulableTaint,
+		Value:  "true",
+		Effect: corev1.TaintEffectNoSchedule,
+	}
+	nn.Spec.Taints = append(nn.Spec.Taints, t)
+	nn, err = pcc.client.CoreV1().Nodes().Update(context.TODO(), nn, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Error(err)
+	}
+}
+
+func (pcc *Controller) deTaintNodeNotSchedulable(name string) {
+	node, err := pcc.nodeLister.Get(name)
+	if err != nil {
+		klog.Error(err)
+		return
+	}
+	taints := node.Spec.Taints
+	taints, deleted := utils.DeleteTaintsByKey(taints, constant.NodeNotSchedulableTaint)
+	if !deleted {
+		klog.Infof("detaint %s: no key %s exists, nothing to do\n", name, constant.NodeNotSchedulableTaint)
+		return
+	}
+	nn := node.DeepCopy()
+	nn.Spec.Taints = taints
+	nn, err = pcc.client.CoreV1().Nodes().Update(context.TODO(), nn, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Error(err)
+	}
+}
+
+func (pcc *Controller) nodeWorker() {
+	for {
+		item, shutdown := pcc.nodeUpdateQueue.Get()
+		if shutdown {
+			klog.Info("node work queue shutdown")
+			return
+		}
+
+		nodeTaint := item.(nodeTaintItem)
+		klog.Infof("get node taint item: %v\n", nodeTaint)
+
+		if nodeTaint.taint {
+			pcc.deTaintNodeNotSchedulable(nodeTaint.nodeName)
+		} else {
+			pcc.deTaintNodeNotSchedulable(nodeTaint.nodeName)
+		}
+
+		pcc.nodeUpdateQueue.Done(item)
+	}
+}
+
+func (pcc *Controller) nodePoolWorker() {
+	for {
+		item, shutdown := pcc.nodePoolUpdateQueue.Get()
+		if shutdown {
+			klog.Info("nodepool work queue shutdown")
+			return
+		}
+
+		nodePoolOp := item.(nodePoolItem)
+		klog.Infof("get nodepool item: %v\n", nodePoolOp)
+
+		if nodePoolOp.add {
+			pcc.nodepoolMap.Add(nodePoolOp.nodepoolName, nodePoolOp.nodeName)
+		} else {
+			pcc.nodepoolMap.Del(nodePoolOp.nodepoolName, nodePoolOp.nodeName)
+		}
+
+		pcc.nodePoolUpdateQueue.Done(item)
+	}
+}
+
+func (pcc *Controller) Run(stopCH <-chan struct{}) {
+	if !cache.WaitForCacheSync(stopCH, pcc.nodeSynced, pcc.leaseSynced) {
+		klog.Error("sync poolcoordinator controller timeout")
+	}
+
+	klog.Info("populate nodepool map")
+	nl, err := pcc.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Error(err)
+	}
+	pcc.nodepoolMap.Sync(nl)
+
+	defer pcc.nodeUpdateQueue.ShutDown()
+	defer pcc.nodePoolUpdateQueue.ShutDown()
+
+	klog.Info("start node taint worker")
+	go pcc.nodeWorker()
+
+	klog.Info("start nodepool maintenance worker")
+	go pcc.nodePoolWorker()
+
+	klog.Info("create webhook")
+	go webhook.Run(pcc.nodeLister, pcc.leaseLister, pcc.nodepoolMap)
```

Comment:
maybe it's hard to understand that run webhook in the controller, how about create an independent webhook server for all webhooks in yurt-controller-manager? so it's easy to add other webhooks in the new server in the future.